src/Extension/FluentSearchFormExtension.php

```diff
@@ -0,0 +1,103 @@
+use SilverStripe\ORM\PaginatedList;
+use SilverStripe\ORM\SS_List;
+use TractorCow\Fluent\State\FluentState;
+
+/**
+ * Simple extension to enable SS Core search to work with SS Fluent
+ */
+class FluentSearchFormExtension extends SearchForm
+{
+    /**
+     * Return dataObjectSet of the results using current request to get info from form.
+     * Wraps around {@link searchEngine()}.
+     *
+     * @return PaginatedList
+     */
+    public function getResults()
+    {
+        $keywords = '';
+        $request = $this->getRequestHandler()->getRequest();
+
+        $keyword_patterns = [
+            [
+                'patterns' => [
+                    '/()("[^()"]+")( and )("[^"()]+")()/i',
+                    '/(^| )([^() ]+)( and )([^ ()]+)( |$)/i',
+                ],
+                'callback' => function ($matches) {
+                    return ' +' . $matches[2] . ' +' . $matches[4] . ' ';
+                },
+            ],
+            [
+                'patterns' => [
+                    '/(^| )( not )("[^"()]+")/i',
+                    '/(^| )( not )([^() ]+)( |$)/i',
+                ],
+                'callback' => function ($matches) {
+                    return ' -' . $matches[3];
+                },
+            ],
+        ];
+        foreach ($keyword_patterns as $k_p) {
+            $keywords = preg_replace_callback(
+                $k_p['patterns'],
+                $k_p['callback'],
+                $this->getSearchQuery() ?: ''
+            );
+        }
+
+        // Generate and query for locale based search results
+        $current_locale = FluentState::singleton()->getLocale();
+        $sitetree_table_localised = "SiteTree_Localised_{$current_locale}";
+        $sitetreetables = '"SiteTree_Live"."Title", "SiteTree_Live"."MenuTitle", "SiteTree_Live"."Content", "SiteTree_Live"."MetaDescription"';
+        $sql = <<<SQL
+            SELECT DISTINCT "SiteTree_Live"."ID", "SiteTree_Live"."Title", MATCH ({$sitetreetables}) AGAINST (?) AS Relevance
+            FROM "SiteTree_Live"
+            LEFT JOIN "SiteTree_Localised" AS "{$sitetree_table_localised}"
+            ON "SiteTree_Live"."ID" = "{$sitetree_table_localised}"."RecordID" AND "{$sitetree_table_localised}"."Locale" = ? AND (("{$sitetree_table_localised}"."ID" IS NOT NULL))
+            WHERE (MATCH ({$sitetreetables}) AGAINST (? IN BOOLEAN MODE) + MATCH ({$sitetreetables}) AGAINST (? IN BOOLEAN MODE) AND "SiteTree_Live"."ShowInSearch" <> 0)
+            ORDER BY Relevance DESC
+SQL;
+        $params = [
+            str_replace(['*', '+', '-'], '', $keywords),
+            $current_locale,
+            $keywords,
+            htmlentities($keywords, ENT_NOQUOTES, 'UTF-8')
+        ];
+        $custom_results = DB::prepared_query($sql, $params);
+
+        // Generate results list
+        $results = new ArrayList();
+        $ids = $custom_results->column('ID');
+        if ($ids) {
+            $pages = SiteTree::get()->filter('ID', $ids);
```

Comment:
I just realised that my feedback before ruined your `ORDER BY RELEVANCE` you had working nicely in the earlier query. You could possibly implement it via a join instead, and you could retain the sort column.

Reply:
If your results are large, you may want to move your pagination limit up into this query. See how `setLimitItems` works in MySQLDatabase; By doing this, even if your result sets is large (10,000 items), you only ever load the items for the current page into the php memory.

`->limit($limit, $offset)` to limit the query, and `$list->setLimitItems(false);` when creating paginated list.

See https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/Connect/MySQLDatabase.php#L292-L298

Reply:
There's trouble in getting the total valid query amount to `->setTotalItems()`. I've thought to use `count($ids)`, but that results in a higher count than the total `$results` count as `$ids` contains records we don't have permission to view.

I thought to slice certain indexes from `$ids` in the if statement to check whether we have permission to view, though that doesn't work when adding a limit to our `SiteTree::get()` which will return 10 or 20 records per load, for example. 

I'm unsure how to get only a page lengths worth of results while also attaining the full number of viewable results without doing a query for all SiteTree recods filtered by id.

Does that make sense?

Reply:
Ok well when you do `$custom_results->column()` that invokes the SQL query and gets the list of ids. Instead of building a query, you could put that into an inner join on the SiteTree::get() query.

E.g.


```php

$results = SiteTree::get()
->addInnerJoin(
    '(' . $sql . ')',
    '"Fulltext"."ID" = "SiteTree"."ID"',
    'Fulltext',
    20,
    $params
)
->sort('"Fulltext"."Relevance" DESC');
```

You'll end up with a $results that matches the full results, sorted by the correct relevance.

Then you can do:

```php
$totalCount = $results->count();
$pagedResults = $results->limit($pageSize, $offset);
```

Reply:
Thanks for the suggestion on `innerJoin()`. I've not used that before when returning DOs via the ORM. This seems to be more concise and easier to read IMO.

Now I've got the following. I tried limiting the results before passing them into the `PaginatedList` object and to not do `->setPageLength` or `setPageStart`, though that messed with the pagination left page numbers displaying incorrectly e.g. get param of start=30 exists which should lead to is displaying as page 4, though it displays as page 1.

How does this look to you?

```
// Generate results list
$pages = SiteTree::get()
    ->innerJoin(
        '(' . $sql . ')',
        '"Fulltext"."ID" = "SiteTree"."ID"',
        'Fulltext',
        20,
        $params
    )
    ->sort('"Fulltext"."Relevance" DESC');
// Filter out non-viewable
foreach ($pages as $page) {
    if (empty($page) || !$page->canView()) {
        continue;
    }
    $results->push($page);
}

$results_total_items = $results->count();

// Paginate results
$results_paged = new PaginatedList($results);
$results_paged->setPageStart($page_start);
$results_paged->setPageLength($page_length);
$results_paged->setTotalItems($results_total_items);

return $results_paged;
```